py/desispec/scripts/proc_joint_fit.py

```diff
@@ -230,84 +230,97 @@ def main(args=None, comm=None):
         #- Track number of commands run and number of errors for exit code
         num_cmd = 0
         num_err = 0
+
+        # Commands to run:
+        # - desi_average_fiberflat for each (camera, lampbox) combo (max 30*4 = 120)
+        # - desi_autocalib_fiberflat for each (b,r,z)
+
+        #- temp directory for averaged flats
+        fiberflatnightfile = findfile('fiberflatnight', args.night, args.expids[0], args.cameras[0])
+        fiberflatdirname = os.path.dirname(fiberflatnightfile)
+        tmpdir = os.path.join(fiberflatdirname, "tmp")
+
+        #- expected inputs and outputs
+        inflats = list()
+        inflats_for_camera = dict()
+        average_flats_for_camera = dict()
+        camera_lampboxes = list()
+        flats_for_arm = dict(b=list(), r=list(), z=list())
+        for camera in args.cameras:
+            inflats_for_camera[camera] = list()
+            for expid in args.expids:
+                filename = findfile('fiberflat', args.night, expid, camera)
+                inflats.append(filename)
+                inflats_for_camera[camera].append(filename)
+
+            average_flats_for_camera[camera] = list()
+            for lampbox in range(4):
+                ofile = os.path.join(tmpdir, f"fiberflatnight-camera-{camera}-lamp-{lampbox}.fits")
+                average_flats_for_camera[camera].append(ofile)
+                camera_lampboxes.append( (camera, lampbox, ofile) )
+                arm = camera[0].lower()
+                flats_for_arm[arm].append(ofile)
+
+        #- rank 0 checks if we have enough inputs
+        enough_inputs = True
+        if rank == 0:
```

Comment:
Is there benefit to all ranks performing lines 239-262 here (generating `camera_lampboxes`, `flats_for_arm`, `inflats_for_camera`) as opposed to broadcasting? Maybe the broadcasting is no more efficient than the minor amount of redundant computation, but I just wanted to point it out.

Reply:
Good question.  When MPI ranks are idle at a broadcast barrier waiting for another rank catch up and send them information, they are pegging their individual core doing repeated checks for whether the data have arrived yet.  So in cases where every rank needs the same calculable info, it isn't faster to have one rank calculate + broadcast and it becomes a matter of style for what you consider more readable/maintainable (which is not a strong feature of the MPI parallelism model in the first place...).  I tend to prefer the style above to avoid extra `if rank == 0...`, `if comm is not None` boilerplate.

However, places where I do calculate+broadcast:
  * if the calculation involves I/O (so that only one rank reads the file instead of N>>1 ranks)
  * if the calculation involves significant memory that could cause ranks to compete for the CPU cache